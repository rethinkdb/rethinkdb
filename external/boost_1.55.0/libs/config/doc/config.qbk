[article Boost.Config
    [quickbook 1.4]
    [authors [Beman Dawes, Vesa Karvonen, John Maddock] ]
    [copyright 2001-2007 Beman Dawes, Vesa Karvonen, John Maddock]
    [category broken compiler workarounds]
    [id config]
    [dirname config]
    [purpose
        Helps boost library developers adapt to compiler idiosyncrasies; not intended for library users.
    ]
    [source-mode c++]
    [license
Distributed under the Boost Software License, Version 1.0.
(See accompanying file LICENSE_1_0.txt or copy at
[@http://www.boost.org/LICENSE_1_0.txt])
    ]
]


[/ Cited Boost resources ]

[def __BOOST_REGRESSION_TEST_DRIVER__ [@../../../../tools/regression/doc/index.html boost regression test driver]]
[def __BOOST_CONFIG_HEADER__ [@../../../../boost/config.hpp <boost/config.hpp>]]
[def __BOOST_CONFIG_USER_HEADER__ [@../../../../boost/config/user.hpp <boost/config/user.hpp>]]
[def __BOOST_CONFIG_SUFFIX_HEADER__ [@../../../../boost/config/user.hpp <boost/config/suffix.hpp>]]
[def __BOOST_CONFIG_DIR__ ['<boost-root>]`/boost/config/`]


[/ Other web resources ]

[def __STL_PORT__ [@http://stlport.sourceforge.net STLport]]
[def __BOOST_TRACKER__ [@https://svn.boost.org/trac/boost/newticket Trac]]
[def __CORE_LANGUAGE_DR337__ [@http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#337 Core Language DR337]]
[def __PRINCIPLES_AND_PATTERNS_ARTICLE__ [@http://www.objectmentor.com/resources/articles/Principles_and_Patterns.pdf following article]]


[/ Icons ]

[def __NOTE__ [$images/note.png]]
[def __ALERT__ [$images/caution.png]]
[def __DETAIL__ [$images/note.png]]
[def __TIP__ [$images/tip.png]]
[def __QUESTION_MARK__ [$images/question.png]]
[def __SPACE__ [$images/space.png]]
[def __GO_TO__ [$images/callouts/R.png]]


[/ Document files ]


[include configuring_boost.qbk]
[include macro_reference.qbk]
[include guidelines.qbk]
[include rationale.qbk]
[include acknowledgements.qbk]





