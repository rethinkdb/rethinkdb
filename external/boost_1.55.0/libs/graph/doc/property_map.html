<HTML>
<!--
     Copyright (c) Jeremy Siek, Lie-Quan Lee, and Andrew Lumsdaine 2000
    
     Distributed under the Boost Software License, Version 1.0.
     (See accompanying file LICENSE_1_0.txt or copy at
     http://www.boost.org/LICENSE_1_0.txt)
  -->
<Head>
<Title>Boost Graph Library: Property Map Traits</Title>
<BODY BGCOLOR="#ffffff" LINK="#0000ee" TEXT="#000000" VLINK="#551a8b" 
        ALINK="#ff0000"> 
<IMG SRC="../../../boost.png" 
     ALT="C++ Boost" width="277" height="86"> 

<BR Clear>

<BR Clear>

<H1><A NAME=""></A>
<pre>
property_map&lt;<a href="./PropertyGraph.html">PropertyGraph</a>, <a href="./PropertyTag.html">PropertyTag</a>&gt;
</pre>
</H1>

This is the traits class that produces the type for a <a
href="../../property_map/doc/property_map.html">property map</a> object
for a particular graph type. The property is specified by the <a
href="./PropertyTag.html">PropertyTag</a> template parameter.  Graph classes
must specialize this traits class to provide their own implementation
for property maps.

<pre>
  template &lt;class Graph, class PropertyTag&gt;
  struct property_map {
    typedef ... type;
    typedef ... const_type;
  };
</pre>

<h3>Where Defined</h3>

<a href="../../../boost/graph/properties.hpp"><tt>boost/graph/properties.hpp</tt></a>

<h3>Associated Types</h3>

<table border>

<tr>
<td>Map Type</td>
<td><TT>type</TT> </td>
<td>
The (mutable) property map type for mapping vertex or edge desriptors of
the graph to their associated property values, where the property is
specified by <tt>PropertyTag</tt>. This type is a <a
href="../../property_map/doc/property_map.html">property map</a>.
</tr>

<tr>
<td>Const Map Type</td>
<td><TT>const_type</TT> </td>
<td>
The (non-mutable) property map type for mapping vertex or edge
desriptors of the graph to their associated property values, where the
property is specified by <tt>PropertyTag</tt>. This type is a <a
href="../../property_map/doc/property_map.html">property map</a>.
</tr>

</table>

<br>
<HR>
<TABLE>
<TR valign=top>
<TD nowrap>Copyright &copy; 2000-2001</TD><TD>
<A HREF="http://www.boost.org/people/jeremy_siek.htm">Jeremy Siek</A>,
Indiana University (<A
HREF="mailto:jsiek@osl.iu.edu">jsiek@osl.iu.edu</A>)<br>
<A HREF="http://www.boost.org/people/liequan_lee.htm">Lie-Quan Lee</A>, Indiana University (<A HREF="mailto:llee@cs.indiana.edu">llee@cs.indiana.edu</A>)<br>
<A HREF="http://www.osl.iu.edu/~lums">Andrew Lumsdaine</A>,
Indiana University (<A
HREF="mailto:lums@osl.iu.edu">lums@osl.iu.edu</A>)
</TD></TR></TABLE>

</BODY>
</HTML> 
