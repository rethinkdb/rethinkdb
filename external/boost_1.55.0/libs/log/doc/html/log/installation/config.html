<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=US-ASCII">
<title>Configuring and building the library</title>
<link rel="stylesheet" href="../../../../../../doc/src/boostbook.css" type="text/css">
<meta name="generator" content="DocBook XSL Stylesheets V1.78.1">
<link rel="home" href="../../index.html" title="Chapter&#160;1.&#160;Boost.Log v2">
<link rel="up" href="../installation.html" title="Installation and compatibility">
<link rel="prev" href="../installation.html" title="Installation and compatibility">
<link rel="next" href="../defs.html" title="Definitions">
</head>
<body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF">
<table cellpadding="2" width="100%"><tr><td valign="top"><img alt="Boost C++ Libraries" width="277" height="86" src="../../../../../../boost.png"></td></tr></table>
<hr>
<div class="spirit-nav">
<a accesskey="p" href="../installation.html"><img src="../../../../../../doc/src/images/prev.png" alt="Prev"></a><a accesskey="u" href="../installation.html"><img src="../../../../../../doc/src/images/up.png" alt="Up"></a><a accesskey="h" href="../../index.html"><img src="../../../../../../doc/src/images/home.png" alt="Home"></a><a accesskey="n" href="../defs.html"><img src="../../../../../../doc/src/images/next.png" alt="Next"></a>
</div>
<div class="section">
<div class="titlepage"><div><div><h3 class="title">
<a name="log.installation.config"></a><a class="link" href="config.html" title="Configuring and building the library">Configuring and building the
      library</a>
</h3></div></div></div>
<p>
        The library has a separately compiled part which should be built as described
        in the <a href="http://www.boost.org/doc/libs/release/more/getting_started/" target="_top">Getting
        Started guide</a>. One thing should be noted, though. If your application
        consists of more than one module (e.g. an exe and one or several dll's) that
        use Boost.Log, the library <span class="underline">must</span> be
        built as a shared object. If you have a single executable or a single module
        that works with Boost.Log, you may build the library as a static library.
      </p>
<p>
        The library supports a number of configuration macros:
      </p>
<div class="table">
<a name="log.installation.config.configuration_macros"></a><p class="title"><b>Table&#160;1.1.&#160;Configuration macros</b></p>
<div class="table-contents"><table class="table" summary="Configuration macros">
<colgroup>
<col>
<col>
</colgroup>
<thead><tr>
<th>
                <p>
                  Macro name
                </p>
              </th>
<th>
                <p>
                  Effect
                </p>
              </th>
</tr></thead>
<tbody>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_DYN_LINK</span></code>
                </p>
              </td>
<td>
                <p>
                  If defined in user code, the library will assume the binary is
                  built as a dynamically loaded library ("dll" or "so").
                  Otherwise it is assumed that the library is built in static mode.
                  This macro must be either defined or not defined for all translation
                  units of user application that uses logging. This macro can help
                  with auto-linking on platforms that support it.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_ALL_DYN_LINK</span></code>
                </p>
              </td>
<td>
                <p>
                  Same as <code class="computeroutput"><span class="identifier">BOOST_LOG_DYN_LINK</span></code>
                  but also affects other Boost libraries the same way.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_NO_THREADS</span></code>
                </p>
              </td>
<td>
                <p>
                  If defined, disables multithreading support. Affects the compilation
                  of both the library and users' code. The macro is automatically
                  defined if no threading support is detected.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_CHAR</span></code>
                </p>
              </td>
<td>
                <p>
                  If defined, disables support for narrow character logging. Affects
                  the compilation of both the library and users' code.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_WCHAR_T</span></code>
                </p>
              </td>
<td>
                <p>
                  If defined, disables support for wide character logging. Affects
                  the compilation of both the library and users' code.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_NO_QUERY_PERFORMANCE_COUNTER</span></code>
                </p>
              </td>
<td>
                <p>
                  This macro is only useful on Windows. It affects the compilation
                  of both the library and users' code. If defined, disables support
                  for the <code class="computeroutput"><span class="identifier">QueryPerformanceCounter</span></code>
                  API in the <code class="computeroutput"><span class="identifier">timer</span></code>
                  attribute. This will result in significantly less accurate time
                  readings. The macro is intended to solve possible problems with
                  earlier revisions of AMD Athlon CPU, described <a href="http://support.microsoft.com/?scid=kb;en-us;895980" target="_top">here</a>
                  and <a href="http://support.microsoft.com/?id=896256" target="_top">here</a>.
                  There are also known chipset hardware failures that may prevent
                  this API from functioning properly (see <a href="http://support.microsoft.com/kb/274323" target="_top">here</a>).
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_USE_NATIVE_SYSLOG</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. If for some reason
                  support for the native SysLog API is not detected automatically,
                  define this macro to forcibly enable it
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_DEFAULT_FACTORIES</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. If defined, the parsers
                  for settings will be built without any default factories for filters
                  and formatters. The user will have to register all attributes in
                  the library before parsing any filters or formatters from strings.
                  This can substantially reduce the binary size.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_SETTINGS_PARSERS</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. If defined, none of
                  the facilities related to the parsers for settings will be built.
                  This can substantially reduce the binary size.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_DEBUG_OUTPUT</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. If defined, the support
                  for debugger output on Windows will not be built.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_EVENT_LOG</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. If defined, the support
                  for Windows event log will not be built. Defining the macro also
                  makes Message Compiler toolset unnecessary.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_WITHOUT_SYSLOG</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. If defined, the support
                  for syslog backend will not be built.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_NO_SHORTHAND_NAMES</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of users' code. If defined, some deprecated
                  shorthand macro names will not be available.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_USE_WINNT6_API</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects the compilation of both the library and users' code. This
                  macro is Windows-specific. If defined, the library makes use of
                  the Windows NT 6 (Vista, Server 2008) and later APIs to generate
                  more efficient code. This macro will also enable some experimental
                  features of the library. Note, however, that the resulting binary
                  will not run on Windows prior to NT 6. In order to use this feature
                  Platform SDK 6.0 or later is required.
                </p>
              </td>
</tr>
<tr>
<td>
                <p>
                  <code class="computeroutput"><span class="identifier">BOOST_LOG_USE_COMPILER_TLS</span></code>
                </p>
              </td>
<td>
                <p>
                  Affects only the compilation of the library. This macro enables
                  support for compiler intrinsics for thread-local storage. Defining
                  it may improve performance of Boost.Log if certain usage limitations
                  are acceptable. See below for more comments.
                </p>
              </td>
</tr>
</tbody>
</table></div>
</div>
<br class="table-break"><p>
        You can define configuration macros in the <code class="computeroutput"><span class="identifier">bjam</span></code>
        command line, like this:
      </p>
<pre class="programlisting">bjam --with-log variant=release define=BOOST_LOG_WITHOUT_EVENT_LOG define=BOOST_LOG_USE_WINNT6_API stage
</pre>
<p>
        However, it may be more convenient to define configuration macros in the
        "boost/config/user.hpp" file in order to automatically define them
        both for the library and user's projects. If none of the options are specified,
        the library will try to support the most comprehensive setup, including support
        for all character types and features available for the target platform.
      </p>
<p>
        The logging library uses several other Boost libraries that require building
        too. These are <a href="http://www.boost.org/doc/libs/release/libs/filesystem/doc/index.htm" target="_top">Boost.Filesystem</a>,
        <a href="http://www.boost.org/doc/libs/release/libs/system/doc/index.html" target="_top">Boost.System</a>,
        <a href="http://www.boost.org/doc/libs/release/doc/html/date_time.html" target="_top">Boost.DateTime</a>
        and <a href="http://www.boost.org/doc/libs/release/doc/html/thread.html" target="_top">Boost.Thread</a>.
        Refer to their documentation for detailed instructions on the building procedure.
      </p>
<p>
        One final thing should be added. The library requires run-time type information
        (RTTI) to be enabled for both the library compilation and user's code compilation.
        Normally, this won't need anything from you except to verify that RTTI support
        is not disabled in your project.
      </p>
<h5>
<a name="log.installation.config.h0"></a>
        <span class="phrase"><a name="log.installation.config.notes_about_compiler_supplied_intrinsics_for_tls"></a></span><a class="link" href="config.html#log.installation.config.notes_about_compiler_supplied_intrinsics_for_tls">Notes
        about compiler-supplied intrinsics for TLS</a>
      </h5>
<p>
        Many widely used compilers support builtin intrinsics for managing thread-local
        storage, which is used in several parts of the library. This feature is also
        included in the C++11 standard. Generally, these intrinsics allow for a much
        more efficient access to the storage than any surrogate implementation, be
        that <a href="http://www.boost.org/doc/libs/release/doc/html/thread.html" target="_top">Boost.Thread</a>
        or even native operating system API. However, this feature has several caveats:
      </p>
<div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; ">
<li class="listitem">
            Some operating systems don't support the use of these intrinsics in case
            if the TLS is defined in a shared library that is dynamically loaded
            during the application run time. These systems include Linux and Windows
            prior to Vista. Windows Vista and later do not have this issue.
          </li>
<li class="listitem">
            The TLS may not be reliably accessed from global constructors and destructors.
            At least MSVC 8.0 on Windows is known to have this problem.
          </li>
</ul></div>
<p>
        The library provides the <code class="computeroutput"><span class="identifier">BOOST_LOG_USE_COMPILER_TLS</span></code>
        configuration macro that allows to enable the use of this feature, which
        will improve the library performance at the cost of these limitations:
      </p>
<div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; ">
<li class="listitem">
            The application executable must be linked with the Boost.Log library.
            It should not be loaded dynamically during run time.
          </li>
<li class="listitem">
            The application must not use logging in global constructors or destructors.
          </li>
</ul></div>
</div>
<table xmlns:rev="http://www.cs.rpi.edu/~gregod/boost/tools/doc/revision" width="100%"><tr>
<td align="left"></td>
<td align="right"><div class="copyright-footer">Copyright &#169; 2007-2013 Andrey Semashev<p>
        Distributed under the Boost Software License, Version 1.0. (See accompanying
        file LICENSE_1_0.txt or copy at <a href="http://www.boost.org/LICENSE_1_0.txt" target="_top">http://www.boost.org/LICENSE_1_0.txt</a>).
      </p>
</div></td>
</tr></table>
<hr>
<div class="spirit-nav">
<a accesskey="p" href="../installation.html"><img src="../../../../../../doc/src/images/prev.png" alt="Prev"></a><a accesskey="u" href="../installation.html"><img src="../../../../../../doc/src/images/up.png" alt="Up"></a><a accesskey="h" href="../../index.html"><img src="../../../../../../doc/src/images/home.png" alt="Home"></a><a accesskey="n" href="../defs.html"><img src="../../../../../../doc/src/images/next.png" alt="Next"></a>
</div>
</body>
</html>
