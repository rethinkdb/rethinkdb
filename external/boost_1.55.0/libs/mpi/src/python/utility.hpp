// (C) Copyright 2006 Douglas Gregor <doug.gregor -at- gmail.com>

// Use, modification and distribution is subject to the Boost Software
// License, Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)

//  Authors: Douglas Gregor
#ifndef BOOST_MPI_PYTHON_UTILITY_HPP
#define BOOST_MPI_PYTHON_UTILITY_HPP

/** @file utility.hpp
 *
 *  This file is a utility header for the Boost.MPI Python bindings.
 */
#include <boost/python.hpp>

namespace boost { namespace mpi { namespace python {

template<typename E>
class translate_exception
{
  explicit translate_exception(boost::python::object type) : type(type) { }

public:
  static void declare(boost::python::object type)
  {
    using boost::python::register_exception_translator;
    register_exception_translator<E>(translate_exception(type));
  }

  void operator()(const E& e) const
  {
    using boost::python::object;
    PyErr_SetObject(type.ptr(), object(e).ptr());
  }

private:
  boost::python::object type;
};

} } } // end namespace boost::mpi::python

#endif // BOOST_MPI_PYTHON_UTILITY_HPP
