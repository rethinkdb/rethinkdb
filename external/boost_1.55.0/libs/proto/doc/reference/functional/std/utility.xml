<?xml version="1.0" encoding="utf-8"?>
<!--
  Copyright 2012 Eric Niebler

  Distributed under the Boost
  Software License, Version 1.0. (See accompanying
  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
  -->
<header name="boost/proto/functional/std/utility.hpp">
  <para>Defines Proto callables <computeroutput><classname>boost::proto::functional::make_pair</classname></computeroutput>,
  <computeroutput><classname>boost::proto::functional::first</classname></computeroutput> and
  <computeroutput><classname>boost::proto::functional::second</classname></computeroutput>.</para>

  <namespace name="boost">
    <namespace name="proto">
      <namespace name="functional">

        <!-- proto::functional::make_pair -->
        <struct name="make_pair">
          <purpose>A <conceptname>PolymorphicFunctionObject</conceptname> type that invokes
            <computeroutput>std::make_pair()</computeroutput> on its arguments.</purpose>
          <description>
            <para>
              A <conceptname>PolymorphicFunctionObject</conceptname> type that invokes
              <computeroutput>std::make_pair()</computeroutput> on its arguments.</para>
          </description>
          <inherit>
            <type><classname>proto::callable</classname></type>
          </inherit>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="First"/>
              <template-type-parameter name="Second"/>
            </template>
            <specialization>
              <template-arg>This(First, Second)</template-arg>
            </specialization>
            <typedef name="type">
              <type>std::pair&lt;
        typename boost::remove_const&lt;typename boost::remove_reference&lt;First&gt;::type&gt;::type
      , typename boost::remove_const&lt;typename boost::remove_reference&lt;Second&gt;::type&gt;::type
    &gt;</type>
            </typedef>
          </struct-specialization>
          <method-group name="public member functions">
            <method name="operator()" cv="const">
              <type>typename std::pair&lt; First, Second &gt;</type>
              <template>
                <template-type-parameter name="First"/>
                <template-type-parameter name="Second"/>
              </template>
              <parameter name="first">
                <paramtype>First const &amp;</paramtype>
              </parameter>
              <parameter name="second">
                <paramtype>Second const &amp;</paramtype>
              </parameter>
              <returns>
                <para><computeroutput>std::make_pair(first, second)</computeroutput></para>
              </returns>
            </method>
          </method-group>
        </struct>

        <!-- proto::functional::first -->
        <struct name="first">
          <purpose>
            A <conceptname>PolymorphicFunctionObject</conceptname> type that returns
            the first element of a <computeroutput>std::pair&lt;&gt;</computeroutput>.
          </purpose>
          <description>
            <para>
              A <conceptname>PolymorphicFunctionObject</conceptname> type that returns
              the first element of a <computeroutput>std::pair&lt;&gt;</computeroutput>.</para>
          </description>
          <inherit><type><classname>proto::callable</classname></type>
          </inherit>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="Pair"/>
            </template>
            <specialization>
              <template-arg>This(Pair)</template-arg>
            </specialization>
            <typedef name="type">
              <type>typename Pair::first_type</type>
            </typedef>
          </struct-specialization>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="Pair"/>
            </template>
            <specialization>
              <template-arg>This(Pair &amp;)</template-arg>
            </specialization>
            <typedef name="type">
              <type>typename Pair::first_type &amp;</type>
            </typedef>
          </struct-specialization>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="Pair"/>
            </template>
            <specialization>
              <template-arg>This(Pair const &amp;)</template-arg>
            </specialization>
            <typedef name="type">
              <type>typename Pair::first_type const &amp;</type>
            </typedef>
          </struct-specialization>
          <method-group name="public member functions">
            <method name="operator()" cv="const">
              <type>typename Pair::first_type &amp;</type>
              <template>
                <template-type-parameter name="Pair"/>
              </template>
              <parameter name="pair">
                <paramtype>Pair &amp;</paramtype>
              </parameter>
              <returns>
                <para>
                  <computeroutput>pair.first</computeroutput>
                </para>
              </returns>
            </method>
            <method name="operator()" cv="const">
              <type>typename Pair::first_type const &amp;</type>
              <template>
                <template-type-parameter name="Pair"/>
              </template>
              <parameter name="pair">
                <paramtype>Pair const &amp;</paramtype>
              </parameter>
              <returns>
                <para>
                  <computeroutput>pair.first</computeroutput>
                </para>
              </returns>
            </method>
          </method-group>
        </struct>

        <!-- proto::functional::second -->
        <struct name="second">
          <purpose>
            A <conceptname>PolymorphicFunctionObject</conceptname> type that returns
            the second element of a <computeroutput>std::pair&lt;&gt;</computeroutput>.
          </purpose>
          <description>
            <para>
              A <conceptname>PolymorphicFunctionObject</conceptname> type that returns
              the second element of a <computeroutput>std::pair&lt;&gt;</computeroutput>.
            </para>
          </description>
          <inherit><type><classname>proto::callable</classname></type></inherit>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="Pair"/>
            </template>
            <specialization>
              <template-arg>This(Pair)</template-arg>
            </specialization>
            <typedef name="type">
              <type>typename Pair::second_type</type>
            </typedef>
          </struct-specialization>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="Pair"/>
            </template>
            <specialization>
              <template-arg>This(Pair &amp;)</template-arg>
            </specialization>
            <typedef name="type">
              <type>typename Pair::second_type &amp;</type>
            </typedef>
          </struct-specialization>
          <struct-specialization name="result">
            <template>
              <template-type-parameter name="This"/>
              <template-type-parameter name="Pair"/>
            </template>
            <specialization>
              <template-arg>This(Pair const &amp;)</template-arg>
            </specialization>
            <typedef name="type">
              <type>typename Pair::second_type const &amp;</type>
            </typedef>
          </struct-specialization>
          <method-group name="public member functions">
            <method name="operator()" cv="const">
              <type>typename Pair::second_type &amp;</type>
              <template>
                <template-type-parameter name="Pair"/>
              </template>
              <parameter name="pair">
                <paramtype>Pair &amp;</paramtype>
              </parameter>
              <returns>
                <para>
                  <computeroutput>pair.second</computeroutput>
                </para>
              </returns>
            </method>
            <method name="operator()" cv="const">
              <type>typename Pair::second_type const &amp;</type>
              <template>
                <template-type-parameter name="Pair"/>
              </template>
              <parameter name="pair">
                <paramtype>Pair const &amp;</paramtype>
              </parameter>
              <returns>
                <para>
                  <computeroutput>pair.second</computeroutput>
                </para>
              </returns>
            </method>
          </method-group>
        </struct>

      </namespace>
    </namespace>
  </namespace>

</header>
