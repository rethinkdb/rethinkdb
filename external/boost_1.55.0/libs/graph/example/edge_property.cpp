//=======================================================================
// Copyright 1997, 1998, 1999, 2000 University of Notre Dame.
// Authors: Andrew Lumsdaine, Lie-Quan Lee, Jeremy G. Siek
//
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//=======================================================================
//
//  Sample output:
//
//   0  --(8, 10)--> 1  
//
//   1  --(12, 20)--> 4 --(12, 40)--> 3 
//        <--(8,10)-- 0 <--(16,20)-- 2  
//   2  --(16, 20)--> 1 
//        <--(16,20)-- 5        
//   3  --(12, 40)--> 6 
//        <--(12,40)-- 1        
//   4  --(12, 20)--> 7 
//        <--(12,20)-- 1        
//   5  --(16, 20)--> 2 
//        <--(16,20)-- 6        
//   6  --(16, 20)--> 5 --(8, 10)--> 8  
//        <--(12,20)-- 7        <--(12,40)-- 3  
//   7  --(12, 20)--> 6 
//        <--(12,20)-- 4        
//   8  
//        <--(8,10)-- 6 
//
//
//   0  --(8, 1)--> 1   
//
//   1  --(12, 2)--> 4  --(12, 3)--> 3  
//        <--(8,1)-- 0  <--(16,4)-- 2   
//   2  --(16, 4)--> 1  
//        <--(16,7)-- 5 
//   3  --(12, 5)--> 6  
//        <--(12,3)-- 1 
//   4  --(12, 6)--> 7  
//        <--(12,2)-- 1 
//   5  --(16, 7)--> 2  
//        <--(16,8)-- 6 
//   6  --(16, 8)--> 5  
//        <--(12,10)-- 7        <--(12,5)-- 3   
//   7  --(12, 10)--> 6 
//        <--(12,6)-- 4 
//   8  
//
        
#include <boost/config.hpp>
#include <iostream>

#include <boost/utility.hpp>
#include <boost/property_map/property_map.hpp>
#include <boost/graph/adjacency_list.hpp>


using namespace boost;
using namespace std;


enum edge_myflow_t { edge_myflow };
enum edge_mycapacity_t { edge_mycapacity };

namespace boost {
  BOOST_INSTALL_PROPERTY(edge, myflow);
  BOOST_INSTALL_PROPERTY(edge, mycapacity);
}


template <class Graph>
void print_network(const Graph& G)
{
  typedef typename boost::graph_traits<Graph>::vertex_iterator    Viter;
  typedef typename boost::graph_traits<Graph>::out_edge_iterator OutEdgeIter;
  typedef typename boost::graph_traits<Graph>::in_edge_iterator InEdgeIter;

  typename property_map<Graph, edge_mycapacity_t>::const_type
    capacity = get(edge_mycapacity, G);
  typename property_map<Graph, edge_myflow_t>::const_type    
    flow = get(edge_myflow, G);

  Viter ui, uiend;
  boost::tie(ui, uiend) = vertices(G);

  for (; ui != uiend; ++ui) {
    OutEdgeIter out, out_end;
    cout << *ui << "\t";

    boost::tie(out, out_end) = out_edges(*ui, G);
    for(; out != out_end; ++out)
      cout << "--(" << capacity[*out] << ", " << flow[*out] << ")--> "
           << target(*out,G) << "\t";

    InEdgeIter in, in_end;
    cout << endl << "\t";
    boost::tie(in, in_end) = in_edges(*ui, G);
    for(; in != in_end; ++in)
      cout << "<--(" << capacity[*in] << "," << flow[*in] << ")-- "
           << source(*in,G) << "\t";

    cout << endl;
  }
}


int main(int , char* [])
{
  typedef property<edge_mycapacity_t, int> Cap;
  typedef property<edge_myflow_t, int, Cap> Flow;
  typedef adjacency_list<vecS, vecS, bidirectionalS, 
     no_property, Flow> Graph;

  const int num_vertices = 9;
  Graph G(num_vertices);

  /*          2<----5 
             /       ^
            /         \
           V           \ 
    0 ---->1---->3----->6--->8
           \           ^
            \         /
             V       /
             4----->7
   */

  add_edge(0, 1, Flow(10, Cap(8)), G);

  add_edge(1, 4, Flow(20, Cap(12)), G);
  add_edge(4, 7, Flow(20, Cap(12)), G);
  add_edge(7, 6, Flow(20, Cap(12)), G);

  add_edge(1, 3, Flow(40, Cap(12)), G);
  add_edge(3, 6, Flow(40, Cap(12)), G);

  add_edge(6, 5, Flow(20, Cap(16)), G);
  add_edge(5, 2, Flow(20, Cap(16)), G);
  add_edge(2, 1, Flow(20, Cap(16)), G);

  add_edge(6, 8, Flow(10, Cap(8)), G);

  typedef boost::graph_traits<Graph>::edge_descriptor Edge;

  print_network(G);

  property_map<Graph, edge_myflow_t>::type
    flow = get(edge_myflow, G);

  boost::graph_traits<Graph>::vertex_iterator v, v_end;
  boost::graph_traits<Graph>::out_edge_iterator e, e_end;
  int f = 0;
  for (boost::tie(v, v_end) = vertices(G); v != v_end; ++v)
    for (boost::tie(e, e_end) = out_edges(*v, G); e != e_end; ++e)
      flow[*e] = ++f;
  cout << endl << endl;

  remove_edge(6, 8, G);

  print_network(G);

          
  return 0;
}
