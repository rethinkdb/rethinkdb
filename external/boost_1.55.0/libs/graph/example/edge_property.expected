0	--(8, 10)--> 1	
	
1	--(12, 20)--> 4	--(12, 40)--> 3	
	<--(8,10)-- 0	<--(16,20)-- 2	
2	--(16, 20)--> 1	
	<--(16,20)-- 5	
3	--(12, 40)--> 6	
	<--(12,40)-- 1	
4	--(12, 20)--> 7	
	<--(12,20)-- 1	
5	--(16, 20)--> 2	
	<--(16,20)-- 6	
6	--(16, 20)--> 5	--(8, 10)--> 8	
	<--(12,20)-- 7	<--(12,40)-- 3	
7	--(12, 20)--> 6	
	<--(12,20)-- 4	
8	
	<--(8,10)-- 6	


0	--(8, 1)--> 1	
	
1	--(12, 2)--> 4	--(12, 3)--> 3	
	<--(8,1)-- 0	<--(16,4)-- 2	
2	--(16, 4)--> 1	
	<--(16,7)-- 5	
3	--(12, 5)--> 6	
	<--(12,3)-- 1	
4	--(12, 6)--> 7	
	<--(12,2)-- 1	
5	--(16, 7)--> 2	
	<--(16,8)-- 6	
6	--(16, 8)--> 5	
	<--(12,10)-- 7	<--(12,5)-- 3	
7	--(12, 10)--> 6	
	<--(12,6)-- 4	
8	
	
